migration-tool/src/main/java/software/amazon/awssdk/migration/internal/recipe/S3StreamingRequestToV2.java

```diff
@@ -0,0 +1,173 @@
+        new MethodMatcher("com.amazonaws.services.s3.AmazonS3 "
+                          + "putObject(java.lang.String, java.lang.String, java.io.File)",
+                          true);
+    private static final JavaType.FullyQualified V1_PUT_OBJECT_REQUEST =
+        TypeUtils.asFullyQualified(JavaType.buildType("com.amazonaws.services.s3.model.PutObjectRequest"));
+    private static final JavaType.FullyQualified REQUEST_BODY =
+        TypeUtils.asFullyQualified(JavaType.buildType("software.amazon.awssdk.core.sync.RequestBody"));
+
+    @Override
+    public String getDisplayName() {
+        return "S3StreamingRequestToV2";
+    }
+
+    @Override
+    public String getDescription() {
+        return "S3StreamingRequestToV2.";
+    }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new Visitor();
+    }
+
+    private static final class Visitor extends JavaIsoVisitor<ExecutionContext> {
+        @Override
+        public J.MethodInvocation visitMethodInvocation(J.MethodInvocation method, ExecutionContext executionContext) {
+            if (PUT_OBJECT.matches(method, false)) {
+                method = transformPutFileOverload(method);
+            }
+            return super.visitMethodInvocation(method, executionContext);
+        }
+
+        private J.MethodInvocation transformPutFileOverload(J.MethodInvocation method) {
+            JavaType.Method methodType = method.getMethodType();
+            if (methodType == null) {
+                return method;
+            }
+
+            List<Expression> originalArgs = method.getArguments();
+
+            Expression bucketExpr = originalArgs.get(0);
+            Expression keyExpr = originalArgs.get(1);
+            Expression fileExpr = originalArgs.get(2);
+
+            List<Expression> newArgs = new ArrayList<>();
+            Expression getObjectExpr = bucketAndKeyToPutObject(bucketExpr, keyExpr);
+            newArgs.add(getObjectExpr);
+
+            Space fileArgPrefix = fileExpr.getPrefix();
+            fileExpr = fileToRequestBody(fileExpr.withPrefix(Space.EMPTY)).withPrefix(fileArgPrefix);
+            newArgs.add(fileExpr);
+
+            List<String> paramNames = Arrays.asList("request", "file");
+            List<JavaType> paramTypes = newArgs.stream()
+                                               .map(Expression::getType)
+                                               .collect(Collectors.toList());
+
+
+            methodType = methodType.withParameterTypes(paramTypes)
+                                   .withParameterNames(paramNames);
+
+            return method.withMethodType(methodType).withArguments(newArgs);
+        }
+
+        private J.MethodInvocation fileToRequestBody(Expression fileExpr) {
+            maybeAddImport(REQUEST_BODY);
+
+            J.Identifier requestBodyId = IdentifierUtils.makeId(REQUEST_BODY.getClassName(), REQUEST_BODY);
+
+            JavaType.Method fromFileType = new JavaType.Method(
```

Comment:
Those 9-argu ctors always look intimating to me 😨

Reply:
Yeah. I think if it becomes a bigger issue we can add some util methods to take care of the boilerplate